src/pymor/algorithms/chol_qr.py

```diff
@@ -0,0 +1,169 @@
+    elif dtype == np.float64:
+        xtrmm, xtrtri = spla.blas.dtrmm, spla.lapack.dtrtri
+    elif dtype == np.complex64:
+        xtrmm, xtrtri = spla.blas.ctrmm, spla.lapack.ctrtri
+    elif dtype == np.complex128:
+        xtrmm, xtrtri = spla.blas.ztrmm, spla.lapack.ztrtri
+
+    iter = 1
+    shift = None
+    eig = None
+    while iter <= maxiter:
+        shift = None
+        with logger.block(f'Iteration {iter}'):
+            # This will compute the Cholesky factor of the lower right block
+            # and keep applying shifts if it breaks down.
+            while True:
+                try:
+                    Rx = spla.cholesky(X - B@B.T, overwrite_a=True, check_finite=check_finite)
+                    break
+                except spla.LinAlgError:
+                    pass
+                logger.warning('Cholesky factorization broke down! Matrix is ill-conditioned.')
+                if shift is None:
+                    m, n = A.dim, len(A[offset:])
+                    if product is None:
+                        shift = m*n+n*(n+1)
+                        XX = X
+                    else:
+                        if not eig:
+                            from pymor.algorithms.eigs import eigs
+                            eig = eigs(product, k=1)[0][0]
+
+                        shift = 2*m*np.sqrt(m*n)+n*(n+1)*np.abs(eig)
+                        XX = A[offset:].gramian(product=product)
+                    shift *= 11*eps*spla.eigh(XX, eigvals_only=True, subset_by_index=[n-1, n-1], driver='evr')[0]**2
+
+                logger.info(f'Applying shift: {shift}')
+                idx = range(len(X))
+                X[idx, idx] += shift
+
+            # orthogonalize
+            Rinv = xtrtri(Rx)[0].T
+            A_todo = A[:offset].lincomb(-Rinv@B) + A[offset:].lincomb(Rinv)
+            del A[offset:]
+            A.append(A_todo)
+
+            # update blocks of R
+            if iter == 1:
+                Bi = B.T
+                Ri = Rx
+            else:
+                Bi += B.T @ Rx
+                xtrmm(1, Rx, Ri, overwrite_b=True)
+
+            # computation not needed in the last iteration
+            if iter < maxiter:
+                B, X = np.split(A[offset:].inner(A, product=product), [offset], axis=1)
+                B = B.conj()
+            elif orth_tol is not None:
+                X = A[offset:].gramian(product=product)
+
+            # check orthonormality (for an iterative algorithm)
+            if orth_tol is not None:
+                res = spla.norm(X - np.eye(len(A) - offset), ord='fro', check_finite=check_finite)
+                logger.info(f'Residual = {res}')
+                if res <= orth_tol*np.sqrt(len(A)):
+                    break
+                elif iter == maxiter:
+                    logger.warning('Orthonormality could not be achieved within the given tolerance. \
+                    Consider increasing maxiter.')
```

Comment:
I believe what we discussed - or at least what I would prefer - is that the algorithm should fail if the orthogonality condition cannot be satisfied after maxiter iterations.  However, looking again at other algorithms in pyMOR, it seems like we don't have an established standard for maxiter #1372.